### ISSUE TITLE
split up events

### ISSUE THREAD

**Issue by marten-seemann** [quicwg/qlog#85]
This is a high-level observation based on my experience of 1. adding qlog support for quic-go and 2. writing some tooling to sift through a pile of logs.

When adding qlog support for quic-go, I already found it confusing that qlog bundles things that have little to do with each other into the same event. Examples for these are:
* The values in `parameters_set` event in `transport` that are not part of the QUIC transport parameters.
* The `pto_count` in `metrics_updated`.
* The `connection_state_updated` event. This is the worst in this list.

In my implementation, I worked around this by exposing an API that exposes different functions, which then would encode the same qlog event (with the respective fields set). For example, I have a `qlog.UpdatedPTOCount()` and a `qlog.UpdatedMetrics()`, which both emit a `metrics_updated` event.

Writing my own qlog parser, I'm now encountering the same problem again. For example, one interesting thing to monitor on a production system would be the reason why a connection was closed (to see if there are any PROTOCOL_VIOLATIONS, for example). It would make filtering much easier if there was a `connection_closed` event, instead of having to look at optional fields in all the `connection_state_updated` events emitted over the lifetime of the connection. A similar argument applies to the PTO count (Connections that experience a lot of PTO events might be interesting to look at. Connections that collect many RTT samples are not).

From my discussions with @rmarx (and I hope that I'm paraphrasing him correctly here), the main argument for having fewer events was that it makes things easier to implement. I hope that I laid out my argument here that the opposite is the case.

**Comment by rmarx** [comment]
So to me, this is a core question that I'd like to resolve before finalizing draft-02.

My viewpoint has indeed always been that too many individual events and definitions would lead people to support only a minor subset. As it is, even with heavy coalescing, we have a large amount of events today and many implementations only implement a subset of those. 

That being said, I can see the appeal of splitting things out again, yet that would mean doing that at all layers (e.g., you can't have everything split out at `transport` but keep everything aggregated in `http` (or am I wrong about that?)).

There is another aspect of "backwards compatibility" with existing implementations that are being re-used (e.g., Lucas' Rust crate, qvis). I'm not sure if people are willing to make major changes at this point. On the other end: we probably shouldn't get bogged down by that at this point yet.

So I'd like some more input on this from other active implementers/contributors: @lpardue, @huitema, @jlaine, @nibanks, @mpiraux

**Comment by rmarx** [comment]
After further discussion on slack, I think @marten-seemann's overall viewpoint is:

- Events can be grouped, but only if they (a) logically belong together and/or (b) share the same fields

This leads to the following concrete example changes:
- In the example of `connection_state_updated`, if there is a `reason` for the `connection_close` (see #78), that reason won't be present for `connection_attempted` and so it makes sense to split up those events.
- In the example of `metrics_updated`, almost everything happens as the result of receiving an ACK, but `pto_count` does not (due to timer firing), so they should be split up as well. (tangent: max_ack_delay is a constant and should probably be in a `parameters_set` recovery equivalent)

Tentative conclusion: we don't need to split up everything as it stands now, just go through the events and make sure they adhere to the general principle above. I tend to agree with that.

**Comment by rmarx** [comment]
As indicated in the last comment above, I understand the arguments behind wanting to split up connection_state_updated and metrics_updated. 

I have now however spent a considerable amount of time thinking about splitting up `transport:parameters_set`, and I just don't see how it can be improved practically without introducing a large amount of one-of events or ending up with 2 very similar events in concept. 

Concretely, if parameters_set would indeed only contain the actual TLS-transport "transport parameters" from the QUIC draft, we'd need either:
1. a separate `settings_set` event (or `configuration_chosen` or something) to contain other things in there (like chosen ALPN value, chosen version, chosen cipher, whether a retry is always required (e.g., DDoS prevention mode), whether early data is supported, etc.)
2. multiple separate events, one for each thing: `version_selected`, `alpn_selected`, `cipher_selected`, ... etc. 

This is then made more complicated by other issues that I can't seem to place easily either:
- #28: how to list which ALPNs an endpoint supports (not the same as the one it actually sends/chooses)
- #75: similar for the versions an endpoint supports (especially needed as not all connections send a VNEG containing (some of) this info)
- #88: indicate which parameters were restored for 0-RTT (in that issue, @marten-seemann even advocates splitting up `parameters_set` into `parameters_sent` and `parameters_restored`, creating even more event types). 

All these things could (for me) conceptually just be part of `parameters_set` (e.g., new `supported_versions` and `chosen_version` fields, logged at appropriate times with the correct owner). 
The other option would be to split this out into several events, ending up with `local_configuration_set`, `configuration_chosen`, `parameters_restored`, etc. next to `parameters_set`... which, tbh, when listing it like that, I'm not sure is less confusing in the long run than having the single `parameters_set`.

I somewhat have the feeling that I don't understand @marten-seemann's motivations for wanting to split up `parameters_set` the same way I do the two other examples in the first comment of this issue. At the moment, it feels mostly because of the semantics of the term `parameters` which is used for an explicit purpose in the QUIC drafts. Put differently: if we just renamed `parameters_set` to `configuration_chosen` (or similar), would you then still advocate for splitting the event? 

If the answer is yes, I'd really like a concrete proposal for how to split them up exactly (mainly: how many different events would need to be created? option 1 or 2 above? and why?). 

Input from others in this is of course also greatly appreciated, as currently it seems mostly me and @marten-seemann having conflicting viewpoints. If we can't resolve this easily by tomorrow, I will punt this to draft-03.

**Comment by marten-seemann** [comment]
> Put differently: if we just renamed parameters_set to configuration_chosen (or similar), would you then still advocate for splitting the event?

Yes, definitely! This is not about the name, but by the confusing created by stuffing unrelated things into the same event. I'm not sure I understand the motivation for keeping the number of events small, event types are cheap (both in terms of specifying them, in terms of implementing them). 
What's not cheap is writing tooling that needs to figure out what actual event a "catch-all" event like `parameters_set` corresponds to. How do I distinguish between TPs restored and TPs taken from a default configuration? Can a tool rely on a `parameters_set` that sets the ALPN doesn't at the same time set QUIC TPs?

**Comment by rmarx** [comment]
I decided to take the middle road for draft-02 to see how it feels.

I have now split out QUIC versions and ALPN to separate events (`version_information` and `alpn_information`) as those were the two most pressing fields in need of treatment (see also their specific issues, e.g., #74, #28). This does make these things easier to grok and log, I admit. However, doing this for other similar stuff (especially things like TLS ciphers, key share groups, signature algo's, ...) will cause many more event types. 

While I agree that event types are cheap in concept, I've always been afraid that adding all these fine-grained events will feel daunting for new qlog implementers or people updating to a new draft or, especially, new tooling implementers (though there haven't been many of those). If most things are in the `parameters_set` event, which they probably want to support anyway, it should feel like less effort to also support additional fields. I'm not saying that's correct in practice, but just to explain why I've been somewhat hesitant to add many new event types. 

I feel that the relatively small scope of qlog in -00 and -01 has contributed to the adoption levels we're seeing, but maybe the argument can be made that now that we have a "critical mass" we can start splitting things out, expecting people to incrementally follow along. It does also probably make it easier to make "dumb" tools that just plot the data instead of those that do considerably more processing/logic in the tools like qvis (which is the type of tool @marten-seemann is most interested in, IIUC). 

In short, I'm keeping this open, as this is a main general design decision going forward.

**Comment by nibanks** [comment]
I think one high level design decision needs to be made here first. Is the format of the logs supposed to be optimized for the performance of implementations generating them or for tools that are processing them? Generally, I believe it should be optimized for the ones generating them. Tools can afford the extra time and complexity because they're not in the production loop.

Assuming there's general agreement with optimizing for the implementations and not the tools, then splitting up events makes sense in my opinion. For most of these things, the instantaneous action/operation that would trigger these events is very often isolated from much of everything else. Therefore trying to force extra packaging or coalescing requires implementations that don't already have it, to manage additional state.

As to the worry that all the individual events will make it feel daunting for new users/implementers of qlog, this is not unfounded, but things should be structured such that generally as many events as possible are optional, but you only get the full benefit out of the tools if you implement them all. In my opinion, tools should be **very** resilient to missing events for several reasons:

1. Not all may be implemented (yet?).
2. Filtering at collection time may be enabled (e.g. for performance reasons).
3. Events may be lost at collection time (log buffer overflows are **very** common).

This obviously makes things much more complicated for the tool, but that's just unfortunately how things are.

**Comment by rmarx** [comment]
Linking this to #107 for future reference, since that also contains a piece of this puzzle.

**Comment by rmarx** [comment]
So this discussion was important at a time, but since the "middle road" detailed above hasn't received much more input. This was also related to a bunch of other issues that have since been closed (including #58 today) and resolved with adding additional events (e.g., `packets_acked` in #107). 

I still understand the general feedback, but it seems that the major issues have been resolved and splitting things up further would take a lot more design work + re-implementation in a lot of stacks, so I'm closing this without action.
